sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,324 @@
+                    total=len(xs_batches),
+                )
+            ) as _:
+                samples: Tensor
+                samples = Parallel(n_jobs=num_workers)(  # pyright: ignore[reportAssignmentType]
+                    delayed(infer_posterior_on_batch)(
+                        xs_batch, posterior, num_posterior_samples
+                    )
+                    for xs_batch in xs_batches
+                )
+        else:
+            pbar = tqdm(
+                total=num_sim_samples,
+                disable=not show_progress_bar,
+                desc=f"Running {num_sim_samples} samples for tarp analysis.",
+            )
+
+            with pbar:
+                samples = []
+                for xs_batch in xs_batches:
+                    samples.append(
+                        infer_posterior_on_batch(
+                            xs_batch, posterior, num_posterior_samples
+                        )
+                    )
+                    pbar.update(infer_batch_size)
+                samples = torch.cat(samples, dim=1)
+
+        return samples
+
+    def check(
+        self,
+        samples: Tensor,
+        theta: Tensor,
+    ) -> Tuple[Tensor, Tensor]:
+        """
+        Estimates coverage with the TARP method.
+        Reference: `Lemos, Coogan et al 2023 <https://arxiv.org/abs/2302.03026>`_
+
+        Args:
+            samples: The predicted parameter samples to compute the coverage of,
+                     with shape ``(n_samples, n_sims, n_dims)``. These are
+                     obtained by sampling a trained posterior.  Multiple
+                     (posterior) samples for one observation are encouraged.
+            theta: The true parameter value theta, with shape ``(n_sims, n_dims)``.
+
+        Returns:
+            ecp: Expected coverage probability (``ecp``)
+            alpha: credibility values
+
+        """
+        # TARP assumes that the predicted thetas are sampled from the "true"
+        # PDF num_samples times
+        theta = theta.detach() if len(theta.shape) != 2 else theta.detach().unsqueeze(0)
+        samples = samples.detach()
+
+        num_samples = samples.shape[0]  # samples per simulation
+        num_sims = samples.shape[-2]
+        num_dims = samples.shape[-1]
+
+        if self.n_bins is None:
+            self.n_bins = num_sims // 10
+
+        if theta.shape[-2] != num_sims:
+            raise ValueError("theta must have the same number of rows as samples")
+        if theta.shape[-1] != num_dims:
+            raise ValueError("theta must have the same number of columns as samples")
+
+        if self.do_norm:
+            lo = torch.min(theta, dim=-2, keepdims=True).values  # min along num_sims
+            hi = torch.max(theta, dim=-2, keepdims=True).values  # max along num_sims
+            samples = (samples - lo) / (hi - lo + 1e-10)
+            theta = (theta - lo) / (hi - lo + 1e-10)
+
+        assert len(theta.shape) == len(samples.shape)
```

Comment:
I am confused by this assert. I assumed that `samples` always has one dimension more than `theta` because it contains samples for many different `x`?

Reply:
True, I overlooked that. Removed with a5519a93.